p2p/socket.go

```diff
@@ -0,0 +1,461 @@
+		// TODO(derrandz): replace with passed down context ?
+		case <-ctx.Done():
+			break reader
+
+		case <-s.runner.Done():
+			break reader
+
+		case <-s.done:
+			break reader
+
+		case _, open := <-s.reading:
+			if !open {
+				break reader
+			}
+
+		default:
+			{
+				buf, n, err := s.readChunk()
+				if err != nil {
+
+					switch err {
+					case io.EOF:
+						s.error(ErrPeerHangUp(err))
+						break reader
+
+					case io.ErrUnexpectedEOF:
+						s.error(ErrPeerHangUp(err))
+						break reader
+
+					default:
+						s.error(ErrUnexpected(err))
+						break reader
+					}
+				}
+
+				if n == 0 {
+					continue
+				}
+
+				nonce, _, data, wrapped, err := s.c.decode(buf)
+				if err != nil {
+					s.error(err)
+					break reader
+				}
+
+				if nonce != 0 {
+					_, ch, found := s.requests.Find(nonce)
+					if !found {
+						// report that we've received a nonced message whose requested does not exist on our end!
+					}
+
+					ch <- types.NewWork(nonce, data, s.addr, wrapped)
+					close(ch)
+					continue
+				}
+
+				s.runner.Sink() <- types.NewWork(nonce, data, s.addr, wrapped)
+			}
+		}
+	}
+}
+
+// TODO(derrandz): add buffered write by writing exactly the writeBufferSize. Will need to think about how to split big payloads and sequence them
+// writes a chunk=writeBufferSize to the writer
+// this operation is blocking, and blocks until a waiter is ready to receive the signal from the write buffer (signaling that is has written)
+// This is used by send/request/broadcast operations
+// upon each send, the write routine will receive a signal so that it may proceed to send the write over the network
+func (s *socket) writeChunk(b []byte, iserroreof bool, reqnum uint32, wrapped bool) (uint, error) {
+	defer s.buffers.write.Unlock()
+	s.buffers.write.Lock()
+
+	writeBuffer := s.buffers.write.Ref()
+
+	buff := s.c.encode(Binary, iserroreof, reqnum, b, wrapped)
+	(*writeBuffer) = append((*writeBuffer), buff...)
+
+	s.buffers.write.Signal()
+	return uint(len(b)), nil // TODO(derrandz): should length be of b or of the encoded b
+}
+
+// writeChunkAckful is a writeChunk that expects to receive an ACK response for the chunk it has written
+// This method will create a request, which is basically a nonce to identify the chunk to write, and a channel on which to receive the response
+// the channel is blocking, thus allowing the 'wait to receive the response' behavior.
+// the `read` routine takes care of identifying incoming responses (_using the nonce_) and redirecting them to the waiting channels of the currently-open requests.
+func (s *socket) writeChunkAckful(b []byte, wrapped bool) (types.Work, error) {
+	request := s.requests.Get()
+	requestNonce := request.Nonce()
+
+	if _, err := s.writeChunk(b, false, requestNonce, wrapped); err != nil {
+		s.requests.Delete(requestNonce)
+		return types.NewWork(requestNonce, nil, "", false), err
+	}
+
+	var response types.Work
+
+	select {
+	case response = <-request.Response():
```

Comment:
Why do we need the `=` here?

Reply:
So that it's returned on receipt. Check the return statement at the end of the function.